### DISCUSSION TITLE
Parsing String parameter in path.

### DISCUSSION THREAD

**Discussion by sean-howard** [pointfreeco/swift-url-routing#2] [Q&A] [answered]
Hey!

Thanks for putting together this library, I'm keen to start using it straight away! However I've hit a roadblock that I've been not been able to resolve whilst digging through the documentation.

In your example scenario if the enum argument is an alphanumeric `String` rather than `Int` which Parse printer would I use in the path?

```swift
enum AppRoute {
  case buyBook(id: String)
}

let appRouter = OneOf {

  // GET /books/:id/buy
  Route(.case(AppRoute.buyBook(id:))) {
    Path { "books"; /* Parse printer??? */ "buy" }
  }
}

```

I've attempted to use `CharacterSet` from [this documentation](https://pointfreeco.github.io/swift-parsing/main/documentation/parsing/characterset/) but I'm repeatedly getting compiler errors. Are you able to advise the correct one to use?

Thanks!

**Comment by stephencelis** [comment, marked as answer]
@sean-howard You can use `Parse(.string)` to peel off an entire path component:

```swift
Route(.case(AppRoute.buyBook(id:))) {
  Path { "books"; Parse(.string); "buy" }
}
```

And if you use a `RawRepresentable` type-safe wrapper for the ID, you can use `.string.representing(BookId.self)`:

```swift
struct BookId: RawRepresentable {
  var rawValue: String
}

enum AppRoute {
  case buyBook(id: BookId)
}

let appRouter = OneOf {

  // GET /books/:id/buy
  Route(.case(AppRoute.buyBook(id:))) {
    Path { "books"; Parse(.string.representing(BookId.self); "buy" }
  }
}
```

**Reply by davdroman** [reply]
Is `Parse(.string)` equivalent to `String.parser()`? 🤔

**Reply by stephencelis** [reply]
@davdroman We try to make sure all `.parser()` types are incremental parsers, so there is no `String.parser()` (though we briefly had one defined on a branch if it sounds familiar).

We have a dedicated parser for consuming the rest of a string, which is the `Rest` parser, so `Parse(.string)` in this case is most equivalent to:

```swift
Rest().map(.string)
```

**Reply by sean-howard** [reply]
Thanks for your speedy response on this @stephencelis! Works perfectly 🤌
